### DISCUSSION TITLE
accept callback in config to provide custom data cascade merging instead of only deep-merge/flat-merge

### DISCUSSION THREAD

**Discussion by Thibaultfq** [11ty/eleventy#4112] [Enhancement Queue]
### Is your feature request related to a problem?
Recently i set up forestry as headless CMS for 11ty, which i now use to generate most of my front matter in content templates.

I thereby use the advantage of [11ty data cascades](https://www.11ty.dev/docs/data-cascade/), providing default data for all my content templates with a directory data file. This default data file (JSON) is perfectly editable in forestry. 

To reduce the workload when creating a new page, I want to overwrite some of the values as i create this new page (i.e. referred to as content template in 11ty docs) which is perfectly possible using 11ty data cascades, provided you use the same key for these properties since the front matter of the content template takes priority over the data from a directory data file. Using the same key, values can be overwritten or deep merged.

So, i created a Front Matter template (which field in the front matter YAML could exist to fill in in forestry) based on my default data JSON file. When i now create a new page, i see al the data fields data i could possibly fill in. However, when i leave them blank, **i would expect them to be omitted from the generated front matter (and also deep checked and performed for field_groups/objects)**. This is not the case, as all keys are generates with blank values (i.e. `''`), thereby overwriting any data that comes after it in the data cascade with just blank values...

I know this is more something i should report in the forestry community, which i already did but without any short term solution (they moved my issue into their private repo). Nonetheless, I think this case might exist besides forestry, using any tool that generates your YML front matter.

### Describe the solution you'd like
I propose a configuration option for the data cascade merging which takes a callback and allows you to fully manipulate your data **before** it merges (i.e. access to all different data sources with possibly their merge priority) and, if viable or potentially interesting, a callback that allows data manipulation to **after merging** (i.e. to the merged data set)

This functionality would also provide a solution to an array of other possible cases where some data manipulation is needed.

possible example:

```javascript
module.exports = function(eleventyConfig) {
   eleventyConfig.setDataMerge((dataCascade)=>{
    //dataCascade is an array (or map with key of source?) of Objects which contain data from one specific source 
    //(e.g., global data, dir data, front matter content template data...) sorted by merge priority
    let mergedArray = ... //use whatever you want when flattening all arrays, such .filter() method etc.
    return  mergedArray

  });
};
```
If we do NOT want to touch upon the order of the data cascade, i propose a Promise is provided as parameter that provides one data source at a time from the data cascade, in the existing order of merging.

### Describe alternatives you've considered
When having no control of your front matter but wanting to manipulate this before you build eleventy, poor solutions remain. The current workaround uses the beforeBuild event and actually manipulates the sources files, **which should be avoided at any cost and is a bad solution imo**. Moreover, this triggers a rebuild of eleventy that has to be anticipated upon to prevent getting stuck in an infinite building loop.

```javascript
const editor = require("front-matter-editor");
const path = require("path");
const glob = require("glob");
let isCleaned = false;

module.exports = function (eleventyConfig) {
  eleventyConfig.setDataDeepMerge(true); //deep merge objects so that default data can be used
  eleventyConfig.setWatchThrottleWaitTime(1000); // in milliseconds

  //remove yml properties that are empty due to forestry cms setting
  eleventyConfig.on("beforeBuild", () => {
    //if this code ran less than 2000 ms ago, do not run it again to avoid infinite loop
    if (!isCleaned) {
      isCleaned = true;
      // Run me before the build starts
      const destPath = path.join(__dirname, "src", "pages", ".");

      glob(__dirname + "/src/pages/*.html", {}, (err, filePaths) => {
        console.log("11ty:beforeBuild - Removing empty yml props from following files:\n:" + filePaths);
        filePaths.forEach((filePath, index) => {
          editor
            .read(filePath)
            .data((data, matter) => {
              matter.data = removeEmptyKeysFromObject(data);//custom helper function
              matter.content = matter.content.trim();
            })
            .save(destPath, null, (err, matter) => {
              if (index === filePaths.length - 1) {
                setTimeout(() => {
                  isCleaned = false;
                }, 2000);
              }
              if (err) {
                console.log("Could not write to " + destPath, err);
              }
            });
        });
      });
    }
  });
}
```

Other options are that every headless CMS tool provides you with certain customisation over what they create.

### Additional context
To illustrate this case even more, here is a draft of the current scenario

#### scenario

- created front matter template for all possible values
- fill in values you want to override (e.g., `profile.name`)
- leave values blank you want to take default values (e.g., `summary` or `profile.quote`)

#### Perceived behaviour

All values are generated with blanks or provided values thereby overwriting default data from the data cascade with `''`
```
---
summary: ''
profile:
  name: 'new value that will overwrite defaults'
  quote: ''
---
```
#### expected behaviour

Only generate data from front matter template in the actual yml front matterthat has an non-blank value from filling in the template in forestry. This removal of properties should be deeply checked. For the blank values, default values will be used from the data cascade as front matters are merged (in 11ty)
```
---
profile:
  name: 'new value that will overwrite defaults'
---
```

This manipulation could be possible using a merge callback in eleventy beforeBuild.

Is this the right proposal of solution to my problem or am I using the wrong approach in providing default data?

**Comment by zachleat** [comment]
Related #3937
